src/vodozemac/ecies.rs

```diff
@@ -0,0 +1,209 @@
+        let result = self
+            .inner
+            .take()
+            .ok_or_else(Self::used_up_error)?
+            .establish_inbound_channel(&message)?;
+
+        let message = String::from_utf8_lossy(&result.message).to_string();
+
+        Ok(InboundCreationResult { message, channel: result.ecies.into() })
+    }
+
+    /// Create an [`EstablishedEcies`] session using the other side's Curve25519
+    /// public key and an initial plaintext.
+    ///
+    /// After the channel has been established, we can encrypt messages to send
+    /// to the other side. The other side uses the initial message to
+    /// establishes the same channel on its side.
+    pub fn establish_outbound_channel(
+        &mut self,
+        public_key: &Curve25519PublicKey,
+        initial_message: &str,
+    ) -> Result<OutboundCreationResult, JsError> {
+        let result = self
+            .inner
+            .take()
+            .ok_or_else(Self::used_up_error)?
+            .establish_outbound_channel(public_key.inner, initial_message.as_bytes())?;
+
+        Ok(OutboundCreationResult {
+            initial_message: result.message.encode(),
+            channel: result.ecies.into(),
+        })
+    }
+}
+
+/// An established ECIES session.
+///
+/// This session can be used to encrypt and decrypt messages between the two
+/// sides of the channel.
+#[derive(Clone)]
+#[wasm_bindgen]
+pub struct EstablishedEcies {
+    inner: Arc<Mutex<ecies::EstablishedEcies>>,
+}
+
+#[wasm_bindgen]
+impl EstablishedEcies {
+    /// Get our [`Curve25519PublicKey`].
+    ///
+    /// This public key needs to be sent to the other side so that it can
+    /// complete the ECIES channel establishment.
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        self.inner.lock().unwrap().public_key().into()
+    }
+
+    /// Encrypt the given plaintext using this [`EstablishedEcies`] session.
+    pub fn encrypt(&mut self, message: &str) -> String {
+        self.inner.lock().unwrap().encrypt(message.as_bytes()).encode()
+    }
+
+    /// Decrypt the given message using this [`EstablishedEcies`] session.
+    pub fn decrypt(&mut self, message: &str) -> Result<String, JsError> {
+        let message = ecies::Message::decode(message)?;
+        let result = self.inner.lock().unwrap().decrypt(&message)?;
+
+        Ok(String::from_utf8_lossy(&result).to_string())
+    }
+
+    /// Get the [`CheckCode`] which uniquely identifies this
+    /// [`EstablishedEcies`] session.
+    ///
+    /// This check code can be used to check that both sides of the session are
+    /// indeed using the same shared secret.
+    pub fn check_code(&self) -> CheckCode {
+        self.inner.lock().unwrap().check_code().into()
+    }
+}
+
+/// A check code that can be used to confirm that two [`EstablishedEcies`]
+/// objects share the same secret. This is supposed to be shared out-of-band to
+/// protect against active MITM attacks.
```

Comment:
nit: please explicit MITM

Reply:
Did so in: 86f43b7f619aaf0d42e252bdb270db8bba22bd9d.